test/tracer/tracer.c

```diff
@@ -77,6 +89,49 @@ void* thread2() {
   return 0;
 }
 
+void check_lw_sign_ext() {
+#ifdef RV64G
+  volatile uint64_t  mem[2]   = { 0 };
+  volatile uint64_t* p        = &mem[0];
+  volatile uint64_t  badneg   = 0xbadbeef1badbeef0;
+  volatile uint64_t  checkneg = 0xffffffffbadbeef0;
+  TRACE_PUSH_ON
+lw_neq:
+  asm volatile( "ld   t3, 0(%1)   \n\t"     // t3 <- badneg
+                "ld   t4, 0(%2)   \n\t"     // t4 <- checkneg
+                "sd   t3, 4(%0)   \n\t"     // store badneg to upper half of mem[0]
+                "lw   t3, 0(%0)   \n\t"     // load mem[0]
+                "beq  t3, t4, _okneg \n\t"  // check
+                ".word 0x0 \n\t"            // fail
+                "_okneg: \n\t"
```

Comment:
This code is fine as-is; but when writing inline assembly with internal labels, it's good to use local labels, such as:
```
"beq  t3, t4, 0f    \n\t"
".word 0x0          \n\t"
"0:                 \n\t"
```
`0f` means "branch to the first label 0 going forwards". `beq t3, t4, 1b` would mean "branch to the first label 1 going backwards". The same numbers can be used multiple times because they bind forwards or backwards to the first matching label/branch and won't conflict if repeated later. This is good when the inline assembly is part of a macro, because it can be called several times and its local labels will be independent of local labels in other macro calls.